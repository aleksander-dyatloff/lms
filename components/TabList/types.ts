import Align from '@interfaces/Align'
import ChangeHandler from '@interfaces/ChangeHandler'
import { FC } from 'react'

namespace TabListComponent {
  export interface Props {
    align?: Align
    as?: keyof HTMLElementTagNameMap
    className?: string
  }

  export interface TabSize {
    width: number
    left: number
  }

  export interface ItemProps {
    value: string
    selectedValue: string
    name?: string
    onChange: ChangeHandler
  }

  export interface MainComponent extends FC<Props> {
    Item: FC<ItemProps>
  }
}

export default TabListComponent
