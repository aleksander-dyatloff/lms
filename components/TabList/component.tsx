import combineClasses from '@utils/combineClasses'
import isBrowser from '@utils/isBrowser'
import { MouseEventHandler, useContext, useEffect, useRef, useState } from 'react'

import TabSizingContext from './contexts'
import Wrapper from './styles'
import TabListComponent from './types'

const TabList: TabListComponent.MainComponent = ({
  as = 'div',
  children,
  ...restProps
}) => {
  const [tabSizing, setTabSizing] = useState<TabListComponent.TabSize | null>()

  return (
    <TabSizingContext.Provider value={setTabSizing}>
      <Wrapper
        as={as}
        {...restProps}
      >
        <div
          className='indicator'
          style={{
            width: `${tabSizing?.width}px`,
            left: `${tabSizing?.left}px`,
          }}
        />
        {children}
      </Wrapper>
    </TabSizingContext.Provider>
  )
}

TabList.Item = ({
  name = '',
  value,
  selectedValue,
  onChange,
  ...restProps
}) => {
  const ref = useRef<HTMLButtonElement>(null)

  const wrapperClasses = combineClasses('item reset', selectedValue === value && 'active')

  const getTabItemSize = (): TabListComponent.TabSize => ({
    width: ref.current?.clientWidth ?? 100,
    left: ref.current?.offsetLeft ?? 0,
  })

  const setTabSizing = useContext(TabSizingContext)

  const handleClick: MouseEventHandler<HTMLElement> = () => {
    onChange({
      name, value,
    })

    setTabSizing(getTabItemSize())
  }

  useEffect(() => {
    if (isBrowser() && selectedValue === value) {
      setTabSizing(getTabItemSize())
    }
  }, [])

  return (
    <button
      ref={ref}
      type='button'
      name={name}
      onClick={handleClick}
      className={wrapperClasses}
      {...restProps}
    />
  )
}

export default TabList
