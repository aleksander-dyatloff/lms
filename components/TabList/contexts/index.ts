import { createContext, Dispatch, SetStateAction } from 'react'

import TabListComponent from '../types'

type SetTabSize = Dispatch<SetStateAction<TabListComponent.TabSize | null | undefined>>

const TabSizingContext = createContext<SetTabSize>(() => {

})

export default TabSizingContext
