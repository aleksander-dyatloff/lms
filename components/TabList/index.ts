import TabList from './component'

export default TabList
