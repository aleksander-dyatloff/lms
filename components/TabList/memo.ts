import { memo } from 'react'

import TabList from './component'

export default memo(TabList)
