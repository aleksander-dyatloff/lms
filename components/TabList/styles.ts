import Element from '@components/Element/styles'
import align from '@styles/mixins/align'
import styled from 'styled-components'

const Wrapper = styled(Element)`
  position: relative;
  display: inline-flex;
  justify-content: space-between;
  align-items: center;
  min-height: 32px;

  ${({ align: alignProp }) => align(alignProp)}

  .item {
    cursor: pointer;
    padding: 0 ${({ theme }) => theme.spacing * 3}px;
    transition-property: opacity;
    transition-duration: 160ms;
    transition-timing-function: ease-out;

    &:not(.active) {
      opacity: 0.7;
    }
  }

  .indicator {
    position: absolute;
    height: 4px;
    border-radius: 4px 4px 0 0;
    background-color: white;
    bottom: 0;
    left: 0;
    right: 0;
    transition-property: width, left;
    transition-duration: 160ms;
    transition-timing-function: ease-out;
  }
`
export default Wrapper
