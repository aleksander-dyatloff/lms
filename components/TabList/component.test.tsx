import 'jest-styled-components'

import { render } from '@testing-library/react'
import { Profiler } from 'react'

import TabList from './component'

describe('TabList component', () => {
  test('should match to snapshot', () => {
    const { container } = render(<TabList />)

    expect(container).toMatchSnapshot()
  })

  test('first render duration should match to specification', () => {
    render(
      <Profiler
        id='TabList'
        onRender={(_, __, renderDuration) => {
          expect(renderDuration).toBeLessThanOrEqual(1.5)
        }}
      >
        <TabList />
      </Profiler>,
    )
  })
})
