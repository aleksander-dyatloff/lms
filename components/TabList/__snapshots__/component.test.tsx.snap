// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabList component should match to snapshot 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 32px;
}

.c0 .item {
  cursor: pointer;
  padding: 0 NaNpx;
  -webkit-transition-property: opacity;
  transition-property: opacity;
  -webkit-transition-duration: 160ms;
  transition-duration: 160ms;
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
}

.c0 .item:not(.active) {
  opacity: 0.7;
}

.c0 .indicator {
  position: absolute;
  height: 3px;
  border-radius: 3px 3px 0 0;
  background-color: white;
  bottom: 0;
  left: 0;
  right: 0;
  -webkit-transition-property: width,left;
  transition-property: width,left;
  -webkit-transition-duration: 160ms;
  transition-duration: 160ms;
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
}

<div>
  <div
    class="c0"
  >
    <div
      class="indicator"
    />
  </div>
</div>
`;
